runner.py

```diff
@@ -0,0 +1,653 @@
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=self.env,
+            universal_newlines=True,
+            cwd=PRTOS_PROJECT_DIR
+        )
+
+        out, err = proc.communicate()
+        if proc.returncode != 0 and handle_error:
+            logging.error(f"Command {' '.join(args)} for {self.target} failed!")
+            logging.error(out)
+            logging.error(err)
+            # Fail at build stage should not happen, exit immediately
+            sys.exit(1)
+        elif proc.returncode != 0:
+            logging.info(f"Command {' '.join(args)} for {self.target} success!")
+
+        return proc.returncode, out, err
+
+    def build(self, syspage=None):
+        if syspage:
+            syspage_with_defaults = list(set(syspage + TargetBuilder.SYSPAGE[self.target]))
+            self.env['SYSPAGE'] = " ".join(syspage_with_defaults)
+
+        self.prebuild_action(syspage)
+
+        logging.info(f"Building {self.env['TARGET']} with syspage: {self.env['SYSPAGE']}")
+        self.run_command(['./docker-build.sh', 'clean', 'core', 'fs', 'image', 'project'])
+
+        self.postbuild_action(syspage)
+
+
+class IA32Builder(TargetBuilder):
+    def __init__(self):
+        super().__init__('ia32-generic')
+
+    def prebuild_action(self, syspage=None, **kwargs):
+        # We run syspage progs from filesystem, set only default syspage
+        self.env['SYSPAGE'] = " ".join(self.SYSPAGE[self.target])
+
+    def postbuild_action(self, syspage=None, **kwargs):
+        if not syspage:
+            return
+
+        logging.info(f"Installing {' '.join(syspage)} to filesystem for {self.target}")
+        build_dir = PRTOS_PROJECT_DIR / f"_build/{self.target}/prog.stripped"
+        syspage_dir = pathlib.Path("/root/syspage")
+
+        self.fs_mkdir(syspage_dir)
+        for prog in syspage:
+            self.fs_install(syspage_dir, build_dir / prog, 0o755)
+
+        self.run_command(['./docker-build.sh', 'image'])
+
+
+class TargetBuilderFactory:
+    @staticmethod
+    def create(target):
+        if target == 'ia32-generic':
+            return IA32Builder()
+
+        return TargetBuilder(target)
+
+
+class YAMLParserError(Exception):
+    def __init__(self, parser, message="Parser error"):
+        message = f"An error occured while parsing {parser.path}:\n" + message
+        super().__init__(message)
+
+
+class YAMLParser:
+    def __init__(self, path):
+        with open(path, 'r') as f_yaml:
+            try:
+                self.config = yaml.safe_load(f_yaml)
+            except yaml.YAMLError as exc:
+                raise YAMLParserError(self, exc) from exc
+
+        self.parsed_tests = []
+        # Current path is /example/test/test.yaml - get rid of test.yaml in path
+        self.path = path.parents[0]
+
+    def parse_test_case(self, config):
+        allowed_keys = ['syspage', 'type', 'harness', 'name', 'path', 'attribute']
+        mandatory_keys = ['syspage', 'name', 'path']
```

Comment:
```suggestion
        allowed_keys = ('syspage', 'type', 'harness', 'name', 'path', 'attribute')
        mandatory_keys = ('syspage', 'name', 'path')
```
`tuple` are faster and take up less  memory.

Reply:
I agree that immutable tuples are betters here. Changed